### PR TITLE
Add go control plane validation for routes, clusters and endpoints

### PR DESCRIPTION
**Pull request by Drifftr**
### Purpose
Add go control plane validation for routes, clusters and endpoints

### Issues
https://github.com/wso2-enterprise/choreo/issues/28819

### Automation tests
 - Unit tests added: Yes/No
 - Integration tests added: Yes/No

### Tested environments

Not Tested

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: adapter/internal/oasparser/envoyconf/http_filters.go [hunk 1]
```diff
@@ -95,6 +95,10 @@ func getRouterHTTPFilter() *hcmv3.HttpFilter {
 		Name:       wellknown.Router,
 		ConfigType: &hcmv3.HttpFilter_TypedConfig{TypedConfig: routeFilterTypedConf},
 	}
+	err = filter.Validate()
+	if err != nil {
+		logger.LoggerOasparser.Error("Error while validating Router HTTP filter configs. ", err)
```
Comment (renuka-fernando) [inline adapter/internal/oasparser/envoyconf/http_filters.go:100]:
```suggestion
		logger.LoggerOasparser.Fatal("Error while validating Router HTTP filter configs. ", err)
```

File: adapter/internal/oasparser/envoyconf/http_filters.go [hunk 2]
```diff
@@ -110,7 +114,10 @@ func getCorsHTTPFilter() *hcmv3.HttpFilter {
 		Name:       wellknown.CORS,
 		ConfigType: &hcmv3.HttpFilter_TypedConfig{TypedConfig: corsFilterTypedConf},
 	}
-
+	err = filter.Validate()
+	if err != nil {
+		logger.LoggerOasparser.Error("Error while validating cors filter configs. ", err)
```
Comment (renuka-fernando) [inline adapter/internal/oasparser/envoyconf/http_filters.go:119]:
```suggestion
		logger.LoggerOasparser.Fatal("Error while validating cors filter configs. ", err)
```

File: adapter/internal/oasparser/envoyconf/http_filters.go [hunk 3]
```diff
@@ -176,6 +183,10 @@ func getRateLimitFilter() *hcmv3.HttpFilter {
 			TypedConfig: ext,
 		},
 	}
+	err2 = rlFilter.Validate()
+	if err2 != nil {
+		logger.LoggerOasparser.Error("Error while validating the rate limit filter.", err2)
```
Comment (renuka-fernando) [inline adapter/internal/oasparser/envoyconf/http_filters.go:188]:
```suggestion
		logger.LoggerOasparser.Fatal("Error while validating the rate limit filter.", err2)
```

File: adapter/internal/oasparser/envoyconf/http_filters.go [hunk 4]
```diff
@@ -210,6 +221,10 @@ func getExtAuthzHTTPFilter() *hcmv3.HttpFilter {
 			TypedConfig: ext,
 		},
 	}
+	err2 = extAuthzFilter.Validate()
+	if err2 != nil {
+		logger.LoggerOasparser.Error("Error while validating the ext authz filter.", err2)
```
Comment (renuka-fernando) [inline adapter/internal/oasparser/envoyconf/http_filters.go:226]:
```suggestion
		logger.LoggerOasparser.Fatal("Error while validating the ext authz filter.", err2)
```

File: adapter/internal/oasparser/envoyconf/http_filters.go [hunk 5]
```diff
@@ -236,6 +251,10 @@ func getLuaFilter() *hcmv3.HttpFilter {
 			TypedConfig: ext,
 		},
 	}
+	err2 = luaFilter.Validate()
+	if err2 != nil {
+		logger.LoggerOasparser.Error("Error while validating the lua filter.", err2)
```
Comment (renuka-fernando) [inline adapter/internal/oasparser/envoyconf/http_filters.go:256]:
```suggestion
		logger.LoggerOasparser.Fatal("Error while validating the lua filter.", err2)
```

File: adapter/internal/oasparser/envoyconf/http_filters.go [hunk 6]
```diff
@@ -254,6 +273,10 @@ func getHTTPLocalRateLimitFilter() *hcmv3.HttpFilter {
 			TypedConfig: marshalledRateLimitConfig,
 		},
 	}
+	err = localRateLimitFilter.Validate()
+	if err != nil {
+		logger.LoggerOasparser.Error("Error while validating the local rate limit filter.", err)
```
Comment (renuka-fernando) [inline adapter/internal/oasparser/envoyconf/http_filters.go:278]:
```suggestion
		logger.LoggerOasparser.Fatal("Error while validating the local rate limit filter.", err)
```

File: adapter/internal/oasparser/envoyconf/http_filters.go [hunk 7]
```diff
@@ -312,6 +335,10 @@ func getMgwWebSocketWASMFilter() *hcmv3.HttpFilter {
 			},
 		},
 	}
+	err = mgwWebSocketFilter.Validate()
+	if err != nil {
+		logger.LoggerOasparser.Error("Error while validating web socket filter.", err)
```
Comment (renuka-fernando) [inline adapter/internal/oasparser/envoyconf/http_filters.go:340]:
```suggestion
		logger.LoggerOasparser.Fatal("Error while validating web socket Wasm filter.", err)
```

File: adapter/internal/oasparser/envoyconf/listener.go [hunk 8]
```diff
@@ -288,6 +291,11 @@ func CreateVirtualHosts(vhostToRouteArrayMap map[string][]*routev3.Route) []*rou
 			}
 			virtualHost.RetryPolicy = commonRetryPolicy
 		}
+    
+    err := virtualHost.Validate()
+		if err != nil {
+			logger.LoggerOasparser.Error("Error while validating virtual host configs. ", err)
+    }
```
Comment (renuka-fernando) [inline adapter/internal/oasparser/envoyconf/listener.go:298]:
```suggestion
		}
```

File: adapter/internal/oasparser/envoyconf/listener.go [hunk 9]
```diff
@@ -238,6 +238,10 @@ func createListeners(conf *config.Config) []*listenerv3.Listener {
 			},
 			},
 		}
+		err = listener.Validate()
+		if err != nil {
+			logger.LoggerOasparser.Error("Error while validating listener configs. ", err)
```
Comment (renuka-fernando) [inline adapter/internal/oasparser/envoyconf/listener.go:243]:
```suggestion
			logger.LoggerOasparser.Fatal("Error while validating listener configs. ", err)
```

File: adapter/internal/oasparser/envoyconf/access_loggers.go [hunk 10]
```diff
@@ -126,6 +126,10 @@ func getFileAccessLogConfigs() *config_access_logv3.AccessLog {
 			TypedConfig: accessLogTypedConf,
 		},
 	}
+	err = accessLog.Validate()
+	if err != nil {
+		logger.LoggerOasparser.Error("Error while validating file access log configs. ", err)
```
[the remark this answers is not in the document]
Comment (Drifftr) [reply adapter/internal/oasparser/envoyconf/access_loggers.go]: Thanks, Fixed.

File: adapter/internal/oasparser/envoyconf/access_loggers.go [hunk 11]
```diff
@@ -165,6 +169,10 @@ func getGRPCAccessLogConfigs(conf *config.Config) *config_access_logv3.AccessLog
 			TypedConfig: accessLogTypedConf,
 		},
 	}
+	err = accessLog.Validate()
+	if err != nil {
+		logger.LoggerOasparser.Error("Error while validating grpc access log configs. ", err)
```
[the remark this answers is not in the document]
Comment (Drifftr) [reply adapter/internal/oasparser/envoyconf/access_loggers.go]: Thanks, Fixed.

File: adapter/internal/oasparser/envoyconf/http_filters.go (hunk 1 above)
Comment (Drifftr) [reply adapter/internal/oasparser/envoyconf/http_filters.go]: Thanks, Fixed.

File: adapter/internal/oasparser/envoyconf/http_filters.go (hunk 3 above)
Comment (Drifftr) [reply adapter/internal/oasparser/envoyconf/http_filters.go]: Thanks, Fixed.

File: adapter/internal/oasparser/envoyconf/http_filters.go (hunk 4 above)
Comment (Drifftr) [reply adapter/internal/oasparser/envoyconf/http_filters.go]: Thanks, Fixed.

File: adapter/internal/oasparser/envoyconf/http_filters.go (hunk 5 above)
Comment (Drifftr) [reply adapter/internal/oasparser/envoyconf/http_filters.go]: Thanks, Fixed.

File: adapter/internal/oasparser/envoyconf/http_filters.go (hunk 6 above)
Comment (Drifftr) [reply adapter/internal/oasparser/envoyconf/http_filters.go]: Thanks, Fixed.

File: adapter/internal/oasparser/envoyconf/http_filters.go (hunk 7 above)
Comment (Drifftr) [reply adapter/internal/oasparser/envoyconf/http_filters.go]: Thanks, Fixed.

File: adapter/internal/oasparser/envoyconf/listener.go (hunk 9 above)
Comment (Drifftr) [reply adapter/internal/oasparser/envoyconf/listener.go]: Thanks, Fixed.

**Review by renuka-fernando — DISMISSED** [review]

**Review by renuka-fernando — APPROVED** [review]

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(EastUS) cluster : dev-deployment-v2 : 20240801.12

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(NorthEU) cluster : dev-deployment-v2 : 20240801.12

**Comment by choreo-cicd** [comment]
[succeeded]  : dev-deployment-v2 : 20240801.12

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(CDP2-WU2) cluster : stage-deployment-v2 : 20240801.3

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(EastUS) cluster : stage-deployment-v2 : 20240801.3

**Comment by choreo-cicd** [comment]
[succeeded] Controlplane cluster : stage-deployment-v2 : 20240801.3

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(NorthEU) cluster : stage-deployment-v2 : 20240801.3

**Comment by choreo-cicd** [comment]
[succeeded] Controlplane cluster : prod-deployment-v2 : 20240805.4

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(CDP2-WU2) cluster : prod-deployment-v2 : 20240805.4

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(EastUS) cluster : prod-deployment-v2 : 20240805.4

**Pull request merged**